src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,118 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROW_VALUE = 1.5;
+    private T[] storage;
+    private int size;
+
+    public ArrayList() {
+        storage = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == storage.length) {
+            growCapacity((int) (size * GROW_VALUE));
+        }
+        storage[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds");
+        }
+        if (size == storage.length) {
+            growCapacity((int) (size * GROW_VALUE));
+        }
+        System.arraycopy(storage, index, storage, index + 1, size - index);
+        storage[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        int requiredCapacity = list.size() + size;
+        if (requiredCapacity > storage.length) {
+            growCapacity(requiredCapacity);
+        }
+        for (int i = 0; i < list.size(); i++) {
+            storage[size + i] = list.get(i);
+        }
+        size += list.size();
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexOutOfBoundException(index);
+        return storage[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndexOutOfBoundException(index);
+        storage[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexOutOfBoundException(index);
+        T element = storage[index];
+        removeElement(index);
+        return element;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = getElementIndex(element);
+        if (index == -1) {
+            throw new NoSuchElementException("No such element");
+        }
+        removeElement(index);
+        return element;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void growCapacity(int increaseSize) {
```

Comment:
let's not pass any parameters to this method, calculate increaseSize directly in this method